File: src/main/java/no/nav/familie/ks/oppslag/oppgave/internal/OppgaveClient.java
package no.nav.familie.ks.oppslag.oppgave.internal;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.micrometer.core.instrument.Counter;
import io.micrometer.core.instrument.Metrics;
import no.nav.familie.http.sts.StsRestClient;
import no.nav.familie.ks.kontrakter.oppgave.Oppgave;
import no.nav.familie.ks.oppslag.oppgave.OppgaveIkkeFunnetException;
import no.nav.familie.log.mdc.MDCConstants;
import no.nav.oppgave.v1.FinnOppgaveResponseDto;
import no.nav.oppgave.v1.OppgaveJsonDto;
import org.slf4j.MDC;
import org.springframework.http.*;
import org.springframework.web.client.RestTemplate;

import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.util.Collections;
import java.util.Objects;

public class OppgaveClient {

    private static final String TEMA = "KON";
    private static final String OPPGAVE_TYPE = "BEH_SAK";
    private static final String X_CORRELATION_ID = "X-Correlation-ID";

    private final Counter returnerteIngenOppgaver = Metrics.counter("oppslag.oppgave.response", "antall.oppgaver", "ingen");
    private final Counter returnerteMerEnnEnOppgave = Metrics.counter("oppslag.oppgave.response", "antall.oppgaver", "flerEnnEn");
    private final URI oppgaveUri;
    private final RestTemplate restTemplate;
    private final StsRestClient stsRestClient;
    private final ObjectMapper objectMapper;
    private final String oppgaveBaseUrl;

    public OppgaveClient(String url,
                         RestTemplate restTemplate,
                         StsRestClient stsRestClient,
                         ObjectMapper objectMapper) {
        this.oppgaveUri = URI.create(url + "/api/v1/oppgaver");
        this.restTemplate = restTemplate;
        this.stsRestClient = stsRestClient;
        this.objectMapper = objectMapper;
        this.oppgaveBaseUrl = url;
    }

    public OppgaveJsonDto finnOppgave(Oppgave request) {
        URI requestUrl = lagRequestUrlMed(oppgaveUri, request.getAktorId(), request.getJournalpostId());
        return requestOppgaveJson(requestUrl);
    }

    public OppgaveJsonDto finnOppgave(String oppgaveId) {
        URI requestUrl = URI.create(oppgaveUri + "/" + oppgaveId);
        return requestOppgaveJson(requestUrl);
    }

    public void oppdaterOppgave(OppgaveJsonDto dto, String beskrivelse) {
        dto.setBeskrivelse(dto.getBeskrivelse() + beskrivelse);
        try {
            putRequest(URI.create(oppgaveUri + "/" + dto.getId()), objectMapper.writeValueAsString(dto), String.class);
        } catch (JsonProcessingException e) {
            throw new RuntimeException("Mapping av OppgaveJsonDto til String feilet.", e);
        }
    }

    public void ping() {
        getRequest(URI.create(String.format("%s/internal/alive", oppgaveBaseUrl)), String.class);
    }

    private URI lagRequestUrlMed(URI oppgaveUri, String aktoerId, String journalpostId) {
        return URI.create(oppgaveUri + String.format("?aktoerId=%s&tema=%s&oppgavetype=%s&journalpostId=%s", aktoerId, TEMA, OPPGAVE_TYPE, journalpostId));
    }

    private OppgaveJsonDto requestOppgaveJson(URI requestUrl) {
        var response = getRequest(requestUrl, FinnOppgaveResponseDto.class);
        if (Objects.requireNonNull(response.getBody()).getOppgaver().isEmpty()) {
            returnerteIngenOppgaver.increment();
            throw new OppgaveIkkeFunnetException("Mislykket finnOppgave request med url: " + requestUrl);
        }
        if (response.getBody().getOppgaver().size()>1) {
            returnerteMerEnnEnOppgave.increment();
        }
        return response.getBody().getOppgaver().get(0);
    }

    private <T> ResponseEntity<T> getRequest(URI uri, Class<T> responseType) {
        HttpHeaders headers = new HttpHeaders();
        headers.setBearerAuth(stsRestClient.getSystemOIDCToken());
        headers.add(X_CORRELATION_ID, MDC.get(MDCConstants.MDC_CALL_ID));

        return restTemplate.exchange(uri, HttpMethod.GET, new HttpEntity<>(headers), responseType);
    }

    private <T> void putRequest(URI uri, String requestBody, Class<T> responseType) {
        HttpHeaders headers = new HttpHeaders();
        headers.setBearerAuth(stsRestClient.getSystemOIDCToken());
        headers.add("Content-Type", "application/json;charset=UTF-8");
        headers.add(X_CORRELATION_ID, MDC.get(MDCConstants.MDC_CALL_ID));

        restTemplate.exchange(uri, HttpMethod.PUT, new HttpEntity<>(requestBody, headers), responseType);
    }
}

File: src/main/java/no/nav/familie/integrasjoner/client/rest/OppgaveClient.kt
package no.nav.familie.integrasjoner.client.rest

import com.fasterxml.jackson.core.JsonProcessingException
import io.micrometer.core.instrument.Metrics
import no.nav.familie.integrasjoner.felles.OppslagException
import no.nav.familie.ks.kontrakter.oppgave.Oppgave
import no.nav.familie.log.mdc.MDCConstants
import no.nav.oppgave.v1.FinnOppgaveResponseDto
import no.nav.oppgave.v1.OppgaveJsonDto
import org.slf4j.MDC
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.*
import org.springframework.stereotype.Component
import org.springframework.web.client.RestOperations
import java.net.URI
import java.util.*

@Component
class OppgaveClient(@Value("\${OPPGAVE_URL}") val url: String,
                    @Qualifier("sts") val restTemplate: RestOperations) {

    private val returnerteIngenOppgaver =
            Metrics.counter("oppslag.oppgave.response", "antall.oppgaver", "ingen")
    private val returnerteMerEnnEnOppgave =
            Metrics.counter("oppslag.oppgave.response", "antall.oppgaver", "flerEnnEn")
    private val oppgaveUri: URI
    private val oppgaveBaseUrl: String
    fun finnOppgave(request: Oppgave): OppgaveJsonDto {
        val requestUrl = lagRequestUrlMed(oppgaveUri, request.aktorId, request.journalpostId)
        return requestOppgaveJson(requestUrl)
    }

    fun finnOppgave(oppgaveId: String): OppgaveJsonDto {
        val requestUrl = URI.create("$oppgaveUri/$oppgaveId")
        return requestOppgaveJson(requestUrl)
    }

    fun oppdaterOppgave(dto: OppgaveJsonDto, beskrivelse: String) {
        dto.beskrivelse = dto.beskrivelse + beskrivelse
        try {
            putRequest(URI.create(oppgaveUri.toString() + "/" + dto.id),
                       objectMapper.writeValueAsString(dto),
                       String::class.java)
        } catch (e: JsonProcessingException) {
            throw RuntimeException("Mapping av OppgaveJsonDto til String feilet.", e)
        }
    }

    fun ping() {
        getRequest(URI.create(String.format("%s/client/alive", oppgaveBaseUrl)),
                   String::class.java)
    }

    private fun lagRequestUrlMed(oppgaveUri: URI, aktoerId: String, journalpostId: String): URI {
        return URI.create(oppgaveUri.toString() + String.format("?aktoerId=%s&tema=%s&oppgavetype=%s&journalpostId=%s",
                                                                aktoerId,
                                                                TEMA,
                                                                OPPGAVE_TYPE,
                                                                journalpostId))
    }

    private fun requestOppgaveJson(requestUrl: URI): OppgaveJsonDto {
        val response =
                getRequest(requestUrl, FinnOppgaveResponseDto::class.java)
        if (Objects.requireNonNull(response.body).oppgaver.isEmpty()) {
            returnerteIngenOppgaver.increment()
            throw OppslagException("Ingen oppgaver funnet for $requestUrl",
                                   "oppgave",
                                   OppslagException.Level.MEDIUM,
                                   HttpStatus.NOT_FOUND)
        }
        if (response.body.oppgaver.size > 1) {
            returnerteMerEnnEnOppgave.increment()
        }
        return response.body.oppgaver[0]
    }

    private fun <T> getRequest(uri: URI, responseType: Class<T>): ResponseEntity<T> {
        val headers = HttpHeaders()
        headers.add(X_CORRELATION_ID, MDC.get(MDCConstants.MDC_CALL_ID))
        return restTemplate.exchange(uri,
                                     HttpMethod.GET,
                                     HttpEntity<Any>(headers),
                                     responseType)
    }

    private fun <T> putRequest(uri: URI, requestBody: String, responseType: Class<T>) {
        val headers = HttpHeaders()
        headers.add("Content-Type", "application/json;charset=UTF-8")
        headers.add(X_CORRELATION_ID, MDC.get(MDCConstants.MDC_CALL_ID))
        restTemplate.exchange(uri,
                              HttpMethod.PUT,
                              HttpEntity(requestBody, headers),
                              responseType)
    }

    companion object {
        private const val TEMA = "KON"
        private const val OPPGAVE_TYPE = "BEH_SAK"
        private const val X_CORRELATION_ID = "X-Correlation-ID"
    }

    init {
        oppgaveUri = URI.create("$url/api/v1/oppgaver")
        oppgaveBaseUrl = url
    }
}